apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIrBuilder.kt

```diff
@@ -0,0 +1,658 @@
+          parentNamedFragmentSelectionKeys = parentNamedFragmentSelectionKeys.run {
+            if (namedFragmentSelectionKey != null) {
+              this + namedFragmentSelectionKey
+            } else this
+          },
+          generateFragmentImplementations = generateFragmentImplementations
+      )
+    }
+    val selectionKeys = setOf(
+        parentSelectionKey,
+        parentSelectionKey + fragmentName,
+    ).plus(
+        parentNamedFragmentSelectionKeys.map { selectionKey ->
+          //TODO figure out why this happens when we have nested fragments defined on the same type condition
+          selectionKey
+              .takeIf { selectionKey.keys.last() == fragmentName }
+              ?: selectionKey + fragmentName
+        }
+    ).run {
+      if (namedFragmentSelectionKey != null) {
+        this + namedFragmentSelectionKey
+      } else this
+    }
+    return GenericFragment(
+        name = fragmentName,
+        typeCondition = fragmentTypeCondition,
+        possibleTypes = fragmentPossibleTypes,
+        description = fragmentDescription,
+        selectionSet = selectionSet,
+        conditions = fragmentConditions,
+        nestedFragments = childInlineFragments + childNamedFragments,
+        selectionKeys = selectionKeys,
+    )
+  }
+
+  /**
+   * Formats fragment implementation name by joining type conditions:
+   * ```
+   *  query TestOperation {
+   *   random {
+   *    ... on Being {
+   *    }
+   *    ... on Human {
+   *    }
+   *  }
+   *}
+   * ```
+   * generated name is going to be `BeingHumanRandom`.
+   */
+  private fun List<GenericFragment>.formatFragmentImplementationName(postfix: String): String {
+    return this
+        .distinctBy { fragment -> fragment.typeCondition }
+        .joinToString(separator = "", postfix = postfix.capitalize()) { fragment ->
+          fragment.typeCondition.name!!.capitalize()
+        }
+  }
+
+  /**
+   * Groups provided list of fragments by intersection of possible types.
+   * ```
+   * query TestOperation {
+   *   random {
+   *     ... on Being {
+   *      ... on Human {
+   *      }
+   *      ... on Wookie {
+   *      }
+   *     }
+   *  }
+   *}
+   * ```
+   * as `Human` and `Wookie` are subtypes of `Being` grouped fragment map is going to be:
+   * ```
+   * [
+   *  [Being, Human]: ["Human"],
+   *  [Being, Wookie]: ["Wookie"]
+   * ]
+   * ```
+   */
+  private fun List<GenericFragment>.groupFragmentsByPossibleTypes()
+      : Map<List<GenericFragment>, List<IntrospectionSchema.TypeRef>> {
+    return this
+        .flatMap { fragment -> fragment.possibleTypes }
+        .toSet()
+        .map { possibleType ->
+          possibleType to this.filter { fragment ->
+            fragment.possibleTypes.contains(possibleType)
+          }
+        }.fold(emptyMap()) { acc, (possibleType, fragments) ->
+          acc + (fragments to (acc[fragments]?.plus(possibleType) ?: listOf(possibleType)))
+        }
```

Comment:
At some point we need to restrict the possible types to the type in scope. Imagine the following. schema:

```
interface Being
interface HasLegs

type Human implements Being, HasLegs
type Droid implements Being
type Table implements HasLegs

query {
  hero {
    ... on HasLegs
  }
}
```

We will end up taking `Table` in consideration in the `fromResponse.when` statement